woocommerce/woocommerce-android
# [Stats] background tasks error description

### Description
Enhanced Error Handling in Background Stats Updates

This PR introduces significant improvements to error handling within the background stats update use cases. Previously, these methods returned a simple boolean value, limiting our ability to understand the root cause of errors.

Key Changes:

- Result<Unit> Return Type: Methods now return a Result<Unit> instead of a boolean. This enables more granular error information to be propagated, providing valuable context for debugging and troubleshooting.
- MultipleErrorsException: A new MultipleErrorsException has been introduced to encapsulate multiple exceptions that may occur during a single operation. This provides a more comprehensive view of error conditions.
-  BackgroundUpdateAnalyticsRepository:  A new repository class that will fetch information from the API and return a Result<Unit> has been introduced, making it possible to track errors when something is wrong while refreshing stats.


### Testing information
TC1
- Apply this patch to trigger the sync task when the app goes to the background

<details>
  <summary>Patch</summary>

```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt	(revision 336b90f1969437a8ff7c2ebbbef4ee8f536a6bf7)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt	(date 1727671097330)
@@ -481,10 +481,6 @@
             UpdateDataOnBackgroundWorker.REFRESH_TIME,
             TimeUnit.HOURS
         )
-            .setInitialDelay(
-                UpdateDataOnBackgroundWorker.REFRESH_TIME,
-                TimeUnit.HOURS
-            )
             .setConstraints(constraints)
             .build()
 

```

</details>

- Check that the task is working as expected (look up for the `background_data_synced` event)

TC2
- Apply this patch to trigger the sync task when the app goes to the background but make it fail

<details>
  <summary>Patch</summary>

```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/background/BackgroundUpdateAnalyticsRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/background/BackgroundUpdateAnalyticsRepository.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/background/BackgroundUpdateAnalyticsRepository.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/background/BackgroundUpdateAnalyticsRepository.kt	(revision 336b90f1969437a8ff7c2ebbbef4ee8f536a6bf7)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/background/BackgroundUpdateAnalyticsRepository.kt	(date 1727671109443)
@@ -51,7 +51,7 @@
         }
 
         val errors = awaitAll(currentResult, previousResult)
-            .filter { it.isFailure }
+            .filter { it.isSuccess }
             .map {
                 it.exceptionOrNull()
                     ?: Exception("${BackgroundUpdateAnalyticsRepository::class.java.name} Unknown error")

```
</details>

- Check that when the sync task fails, the errors with the error description information are tracked

### The tests that have been performed
- Checked that the happy path keeps working as expected
- Checked that on a failure, the error description is tracked


## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:
